[WIP] Add and remove servers from upstreams

wfarr opened the pull request:
On the path towards dynamic upstreams.

Heavily WIP. Do not merge.

cc @csfrancis

wfarr committed:
A little bit of README Driven Development

wfarr committed:
Spike out upstream.add_server_to_upstream()

This is based heavily on the work in https://github.com/openresty/lua-upstream-nginx-module/pull/13
but involves slightly fewer changes to the module as a whole.

wfarr committed:
minor fixes

wfarr committed:
test things

csfrancis commented on README.md:
```diff
@@ -12,6 +12,8 @@ Table of Contents
 * [Functions](#functions)
     * [get_upstreams](#get_upstreams)
     * [get_servers](#get_servers)
+    * [add_server_to_upstream](#add_server_to_upstream)
```
I'd call them `add_upstream_peer` and `remove_upstream_peer`.

wfarr committed:
get_servers is a function in upstream module

wfarr committed:
typo

wfarr committed:
test update

wfarr committed:
typo

wfarr committed:
extra newline

wfarr committed:
correct retval

wfarr committed:
rename add_server_to_upstream to add_peer

wfarr committed:
Rename add_peer to add_upstream_peer

wfarr committed:
suspecting race conditions

wfarr committed:
accurate debug logs

wfarr committed:
try %d for bools

wfarr committed:
ok is probably for chumps

wfarr commented:
Tests run on Circle but there appears to be some race condition or double insert occurring that does not make sense to me right now. Will pick this back up again after the weekend.

csfrancis commented on src/ngx_http_lua_upstream_module.c:
```diff
@@ -141,6 +148,92 @@ ngx_http_lua_upstream_get_upstreams(lua_State * L)
+    time_t                       fail_timeout;
+    u_char                       *p;
+
+    if (lua_gettop(L) != 6) {
```
There should be some sane defaults here.

csfrancis commented on src/ngx_http_lua_upstream_module.c:
```diff
@@ -141,6 +148,92 @@ ngx_http_lua_upstream_get_upstreams(lua_State * L)
+
+    host.data = (u_char *) luaL_checklstring(L, 1, &host.len);
+
+    ngx_memzero(&u, sizeof(ngx_url_t));
```
You can declare `u` like so: `ngx_url_t                    u = { 0 };` to automatically zero it.

csfrancis commented on src/ngx_http_lua_upstream_module.c:
```diff
@@ -551,6 +644,32 @@ ngx_http_lua_upstream_find_upstream(lua_State *L, ngx_str_t *host)
 }
 
+static ngx_http_upstream_server_t*
+ngx_http_lua_upstream_find_server(ngx_http_upstream_srv_conf_t * us, ngx_url_t u)
```
The `u` parameters should be a pointer.

wfarr committed:
ngx_http_lua_upstream_find_server should take ngx_url_t pointer

wfarr committed:
Provide default values for upstream.add_upstream_peer()

Only the upstream and HOST:IP are required. For everything else,
we provide sane defaults and treat those arguments as optional.

wfarr committed:
method header fixup

wfarr committed:
misc

wfarr committed:
update test to check primary peers over servers

wfarr committed:
misc

wfarr committed:
attempt to 0 out u correctly

wfarr committed:
Add peers, not servers

wfarr committed:
fixup

wfarr committed:
should be a ptr

wfarr committed:
p is gone

wfarr committed:
remove unused var

wfarr committed:
say error in test

wfarr committed:
better wording

wfarr committed:
lolidunno

wfarr committed:
remove find_server
